sys/kern/cred_checks.c

```diff
@@ -74,3 +75,47 @@ bool cred_can_setlogin(cred_t *cred) {
   /* Only root can setlogin(). */
   return cred->cr_euid == 0;
 }
+
+int cred_can_access(vattr_t *va, cred_t *cred, accmode_t mode) {
+  accmode_t granted = 0;
+  if (cred->cr_euid == va->va_uid) {
+    granted |= VADMIN;
+    if (va->va_mode & S_IRUSR)
+      granted |= VREAD;
+    if (va->va_mode & S_IWUSR)
+      granted |= VWRITE;
+    if (va->va_mode & S_IXUSR)
+      granted |= VEXEC;
+    goto check;
+  }
+
+  if (cred_groupmember(va->va_gid, cred)) {
+    if (va->va_mode & S_IRGRP)
+      granted |= VREAD;
+    if (va->va_mode & S_IWGRP)
+      granted |= VWRITE;
+    if (va->va_mode & S_IXGRP)
+      granted |= VEXEC;
+    goto check;
+  }
+
+  if (va->va_mode & S_IROTH)
+    granted |= VREAD;
+  if (va->va_mode & S_IWOTH)
+    granted |= VWRITE;
+  if (va->va_mode & S_IXOTH)
+    granted |= VEXEC;
+
+check:
+  if (cred->cr_euid == 0) {
+    granted |= VREAD | VWRITE;
+    /* root has exec permission when:
+     *   - file is directory
+     *   - at lest one exec bit is se
+     */
+    if (S_ISDIR(va->va_mode) || va->va_mode & (S_IXUSR | S_IXGRP | S_IXOTH))
+      granted |= VEXEC;
+  }
```

Comment:
Minor thing: I would move this root check to the very beginning of the function, so that the code at the `check` label does as little as possible. I think this would improve readability.